writer.go

```diff
@@ -19,7 +19,25 @@ type Writer struct {
 	framer    Framer
 	formatter Formatter
 
-	conn serverConn
+	conn          serverConn
+	needReconnect bool
+}
+
+// Poll the socket to find out if/when it closes
+func (w *Writer) poll() {
+	for {
+		if w.conn == nil {
+			w.needReconnect = true
+			return
+		}
+
+		b := make([]byte, 1)
+		_, err := w.conn.read(b)
+		if err != nil {
+			w.needReconnect = true
```

Comment:
If we're calling this in a goroutine, we need to protect its memory with a mutex. (Otherwise we're going to get a race condition on reading `w.needReconnect` as false when it's been set to true in a different goroutine.)